sbin/mount/mount.8
.\"	$NetBSD: mount.8,v 1.96 2026/08/26 13:13:44 kre Exp $
.\"
.\" Copyright (c) 1980, 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)mount.8	8.8 (Berkeley) 6/16/94
.\"
.Dd August 26, 2026
.Dt MOUNT 8
.Os
.Sh NAME
.Nm mount
.Nd mount file systems
.Sh SYNOPSIS
.Nm
.Op Fl AadfruvWw
.Op Fl t Ar type
.Nm
.Op Fl dfruvw
.Brq Ar special | Ar node
.Nm
.Op Fl dfruvw
.Op Fl o Ar options
.Op Fl t Ar type
.Ar special node
.Sh DESCRIPTION
The
.Nm
command invokes a file system-specific program to prepare and graft the
.Ar special
device on to the file system tree at the point
.Ar node ,
or to update options for an already-mounted file system.
.Pp
These programs are named
.Dq mount_<fstype>
where
.Dq <fstype>
refers to the file system-specific type; for example
for NFS the mount program is called
.Xr mount_nfs 8 .
.Pp
The
.Ar node
argument is always interpreted as a directory in the name space of
currently mounted file systems.
The
.Ar special
argument is interpreted in different ways
by the programs that handle different file system types;
for example,
.Xr mount_ffs 8
interprets it as a device node,
.Xr mount_null 8
interprets it as a directory name,
.Xr mount_nfs 8
interprets it as reference to a remote host and a directory on that host,
and
.Xr mount_tmpfs 8
ignores it.
.Pp
The system maintains a list of currently mounted file systems.
This list is printed if
.Nm
is invoked with no arguments, and with no options that require some
other behaviour.
.Pp
If exactly one of
.Ar special
or
.Ar node
is provided, then the missing information
(including the file system type)
is taken from the
.Xr fstab 5
file.
The provided argument is looked up first in the
.Dq fs_file ,
then in the
.Dq fs_spec
column.
If the matching entry in
.Xr fstab 5
has the string
.Dq Li from_mount
as its
.Dq fs_spec
field, the device or remote file system already mounted at
the location specified by
.Dq fs_spec
will be used.
.\" XXX The above paragraph doesn't address the use of "-u"
.Pp
If both
.Ar special
and
.Ar node
are provided, then
.Xr fstab 5
is not used.
In this case, if the file system type is not specified
via the
.Fl t
flag, then
.Nm
may determine the type from the disk label (see
.Xr disklabel 8 ) .
In addition, if
.Ar special
contains a colon
.Pq Ql \&:
or at sign
.Pq Ql \&@ ,
then the
.Li nfs
type is inferred.
Otherwise, a
.Xr dk 4
wedge may be referred to by its name with a
.Ar special
argument of the form
.Dq NAME=<value> .
.Pp
In
.Nx ,
the file system mounting policy is dictated by the running security models.
The default security model may allow unprivileged mounting; see
.Xr secmodel_suser 9
and
.Xr secmodel_extensions 9
for details.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl A
Causes
.Nm
to try to mount all of the file systems listed in the
.Xr fstab 5
file except those for which the
.Dq noauto
option is specified.
.It Fl a
Similar to the
.Fl A
flag, except that if a file system (other than the root file system)
appears to be already mounted,
.Nm
will not try to mount it again.
.Nm
assumes that a file system is already mounted if a file system with
the same type is mounted on the given mount point.
More stringent checks are not possible because some file system types
report strange values for the mounted-from device for mounted file
systems.
.It Fl d
Causes everything to be done except for the invocation of
the file system-specific program.
Instead what would have been invoked is printed.
.It Fl f
Forces the revocation of write access when trying to downgrade
a file system mount status from read-write to read-only.
.It Fl o
Options are specified with a
.Fl o
flag followed by a comma separated string of options.
The following options are available:
.Bl -tag -width nocoredump
.It Cm async
All I/O to the file system should be done asynchronously.
In the event of a crash,
.Em "it is impossible for the system to verify the integrity of data on a file system mounted with this option" .
You should only use this option if you have an application-specific data
recovery mechanism, or are willing to recreate the file system from scratch.
.It Cm noasync
Clear
.Cm async
mode.
.It Cm automounted
This flag indicates that the file system was mounted by
.Xr automountd 8 .
Automounted file systems are automatically unmounted by
.Xr autounmountd 8 .
.It Cm discard
Use DISCARD/TRIM commands if disk and driver support it.
.Pp
.Em EXPERIMENTAL
- negatively influences filesystem performance by increasing fragmentation,
causes free block map inconsistency on unclean shutdown,
and is incompatible with
.Cm log .
If
.Cm log
option is also used,

doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.3290 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 11.0 to NetBSD 12.0:

	macppc: Add drivebay(4), a driver to control the Xserve G4's drive bays
		via sysctl. [macallan 20250812]
	macppc: Added support for fan control on Xserve G4. [macallan 20250820]
	acpi(4): Updated ACPICA to 20250807. [christos 20250824]
	binutils: Import 2.45 and switch amd64 to it  [christos 20250826]
	gffb(4): Add backlight control to the GeForce framebuffer driver.
		[macallan 20250830]
	gdb: Import 16.3 and switch everyone to it [christos 20250902]
	openpam(3): update to 20250531 (zingiber) [christos 20250903]
	openldap: Import 2.6.10. [christos 20250904]
	unbound(8): Import 1.23.1. [christos 20250906]
	kernel: Add new memory barrier for paravirtualized drivers
		requiring store-before-load ordering. [riastradh 20250906]
	macppc: Add driver for the LED bar found on Xserve G4,
		controls lights depending on CPU usage. [macallan 20250908]
	m68k: Change softfloat build to libc/softfloat. [nat 20250910]
	fsck_ffs(8): Expand flags with -a and -i. [perseant 20250914]
	dumplfs(8): Add -n flag to allow dumping of a given inode from an
		inode block, or every inode in a block. [perseant 20250914]
	lfs: Add working in-kernel roll forward. [perseant 20250917]
	kernel: Add a shared function to query the common properties used
		for configuring an Ethernet address. [thorpej 20251004]
	gffb(4): Add backlight control for NVIDIA GeForce FX 5200 Go.
		[macallan 20251006]
	hypotl(3): Accommodate infinity for softfloat. [nat 20251007]
	arm, aarch64: Switch to GCC 14. [mrg 20251007]
	hppa: Switch to GCC 14. [mrg 20251007]
	riscv: Switch to GCC 14. [mrg 20251007]
	sparc, sparc64: Switch to GCC 14. [mrg 20251007]
	vax: Switch to GCC 14. [mrg 20251007]
	x86: Switch to GCC 14. [mrg 20251007]
	ucom(4): Add support for console over USB-to-serial adapter
		[manu 20251009]
	libc: Merge extended worker threads for POSIX asynchronous I/O
		(GSoC project by Ethan Miller). [christos 20251010]
	ssh(1), sshd(8): Import OpenSSH 10.2. [christos 20251011]

sbin/mount/mount.c
/*	$NetBSD: mount.c,v 1.109 2026/08/26 13:08:57 kre Exp $	*/

/*
 * Copyright (c) 1980, 1989, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1980, 1989, 1993, 1994\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)mount.c	8.25 (Berkeley) 5/8/95";
#else
__RCSID("$NetBSD: mount.c,v 1.109 2026/08/26 13:08:57 kre Exp $");
#endif
#endif /* not lint */

#include <sys/param.h>
#include <sys/mount.h>
#include <sys/wait.h>

#include <fs/puffs/puffs_msgif.h>

#include <err.h>
#include <errno.h>
#include <fstab.h>
#include <pwd.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <util.h>

#define MOUNTNAMES
#include <fcntl.h>
#include <sys/disk.h>
#include <sys/disklabel.h>
#include <sys/ioctl.h>

#include "pathnames.h"
#include "mountprog.h"

static int	debug, verbose, noise;

static void	catopt(char **, const char *);
static const char *
		getfslab(const char *str);
static struct statvfs *
		getmntpt(const char *);
static int 	getmntargs(struct statvfs *, char *, size_t);
static int	hasopt(const char *, const char *);
static void	mangle(char *, int *, const char ** volatile *, int *);
static int	mountfs(const char *, const char *, const char *,
		    int, const char *, const char *, int, char *, size_t);
static void	prmount(struct statvfs *, int);
__dead static void	usage(void);


/* Map from mount options to printable formats. */
static const struct opt {
	int o_opt;
	int o_silent;
	const char *o_name;
} optnames[] = {
	__MNT_FLAGS
};

#define FLG_UPDATE	1
#define FLG_RDONLY	2
#define FLG_RDWRITE	4
#define FLG_FORCE	8

static const char ffs_fstype[] = "ffs";

int
main(int argc, char *argv[])
{
	const char *mntfromname, *mntonname, **vfslist, *vfstype;
	struct fstab *fs;
	struct statvfs *mntbuf;
#if 0
	FILE *mountdfp;
#endif
	int all, ch, forceall, i, init_flags, mntsize, rval, Wedges;
	char *options;
	const char *mountopts, *fstypename;
	char canonical_path_buf[MAXPATHLEN], buf[MAXPATHLEN];
	char *canonical_path;

	/* started as "mount" */
	all = forceall = init_flags = Wedges = 0;
	options = NULL;
	vfslist = NULL;
	vfstype = ffs_fstype;
	while ((ch = getopt(argc, argv, "Aadfo:rwt:uvW")) != -1)
		switch (ch) {
		case 'A':
			all = forceall = 1;
			break;
		case 'a':
			all = 1;
			break;
		case 'd':
			debug = 1;
			break;
		case 'f':
			init_flags |= FLG_FORCE;
			break;
		case 'o':
			if (*optarg)
				catopt(&options, optarg);
			break;
		case 'r':
			init_flags |= FLG_RDONLY;
			break;
		case 't':
			if (vfslist != NULL)
				errx(EXIT_FAILURE,
				    "Only one -t option may be specified.");
			vfslist = makevfslist(optarg);
			vfstype = optarg;
			break;
		case 'u':
			init_flags |= FLG_UPDATE;
			break;
		case 'v':
			verbose++;
			break;
		case 'W':
			Wedges++;
			break;
		case 'w':
			init_flags |= FLG_RDWRITE;
			break;
		case '?':
		default:
			usage();
			/* NOTREACHED */
		}
	argc -= optind;
	argv += optind;

#define	BADTYPE(type)							\
	(strcmp(type, FSTAB_RO) &&					\
	    strcmp(type, FSTAB_RW) && strcmp(type, FSTAB_RQ))

	noise = 0;
	if (argc > 0 && verbose) {
		noise = 1;
		verbose--;
	}

	rval = 0;
	switch (argc) {
	case 0:
		if (all) {
			while ((fs = getfsent()) != NULL) {
				if (BADTYPE(fs->fs_type))
					continue;
				if (checkvfsname(fs->fs_vfstype, vfslist))
					continue;
				if (hasopt(fs->fs_mntops, "noauto"))
					continue;
				if (strcmp(fs->fs_spec, "from_mount") == 0) {
					if ((mntbuf = getmntpt(fs->fs_file))
					    == NULL)
						errx(EXIT_FAILURE,
						    "Unknown file system %s",
						    fs->fs_file);
					mntfromname = mntbuf->f_mntfromname;
				} else
					mntfromname = fs->fs_spec;
				mntfromname = getfsspecname(buf, sizeof(buf),
				    mntfromname);
				if (mntfromname == NULL)
					err(EXIT_FAILURE, "%s", buf);
				if (mountfs(fs->fs_vfstype, mntfromname,
				    fs->fs_file, init_flags, options,
				    fs->fs_mntops, !forceall, NULL, 0))
					rval = 1;
			}
		} else {
			if ((mntsize = getmntinfo(&mntbuf, MNT_NOWAIT)) == 0)
				err(EXIT_FAILURE, "getmntinfo");
			for (i = 0; i < mntsize; i++) {
				if (checkvfsname(mntbuf[i].f_fstypename,
				    vfslist))
					continue;
				prmount(&mntbuf[i], Wedges);
			}
		}
		return rval;
	case 1:
		if (vfslist != NULL) {
			usage();
			/* NOTREACHED */
		}

		/*
		 * Create a canonical version of the device or mount path
		 * passed to us.  It's ok for this to fail.  It's also ok
		 * for the result to be exactly the same as the original.
		 */
		canonical_path = realpath(*argv, canonical_path_buf);

		if (init_flags & FLG_UPDATE) {
			/*
			 * Try looking up the canonical path first,
			 * then try exactly what the user entered.
			 */
			if ((canonical_path == NULL ||
			    (mntbuf = getmntpt(canonical_path)) == NULL) &&
			    (mntbuf = getmntpt(*argv)) == NULL) {
out:
				errx(EXIT_FAILURE,
				    "Unknown special file or file system `%s'",
				    *argv);
			}
			mntfromname = mntbuf->f_mntfromname;
			if ((fs = getfsfile(mntbuf->f_mntonname)) != NULL) {
				if (strcmp(fs->fs_spec, "from_mount") != 0)
					mntfromname = fs->fs_spec;
				/* ignore the fstab file options.  */
				fs->fs_mntops = NULL;
			}
			mntonname  = mntbuf->f_mntonname;
			fstypename = mntbuf->f_fstypename;
			mountopts  = NULL;
		} else {
			/*
			 * Try looking up the canonical path first,
			 * then try exactly what the user entered.
			 */
			if (canonical_path == NULL ||
			    ((fs = getfsfile(canonical_path)) == NULL &&
			     (fs = getfsspec(canonical_path)) == NULL))
			{
				if ((fs = getfsfile(*argv)) == NULL &&
				    (fs = getfsspec(*argv)) == NULL) {
					goto out;
				}
			}
			if (BADTYPE(fs->fs_type))
				errx(EXIT_FAILURE,
				    "Unknown file system type for `%s'",
				    *argv);
			if (strcmp(fs->fs_spec, "from_mount") == 0) {
				if ((canonical_path == NULL ||
				    (mntbuf = getmntpt(canonical_path))
				     == NULL) &&
				    (mntbuf = getmntpt(*argv)) == NULL)
					goto out;
				mntfromname = mntbuf->f_mntfromname;
			} else
				mntfromname = fs->fs_spec;
			mntonname   = fs->fs_file;
			fstypename  = fs->fs_vfstype;
			mountopts   = fs->fs_mntops;
		}
		mntfromname = getfsspecname(buf, sizeof(buf), mntfromname);
		if (mntfromname == NULL)
			err(EXIT_FAILURE, "%s", buf);
		rval = mountfs(fstypename, mntfromname,
		    mntonname, init_flags, options, mountopts, 0, NULL, 0);
		break;
	case 2:
		/*
		 * If -t flag has not been specified, and spec contains either
		 * a ':' or a '@' then assume that an NFS filesystem is being
		 * specified ala Sun.
		 */
		mntfromname = getfsspecname(buf, sizeof(buf), argv[0]);
		if (mntfromname == NULL)
			err(EXIT_FAILURE, "%s", buf);
		if (vfslist == NULL) {
			if (strpbrk(argv[0], ":@") != NULL) {
				vfstype = "nfs";
			} else {
				vfstype = getfslab(mntfromname);
				if (vfstype == NULL)
					vfstype = ffs_fstype;
			}
		}
		rval = mountfs(vfstype, mntfromname, argv[1], init_flags,
		    options, NULL, 0, NULL, 0);
		break;
	default:
		usage();
	}

#if 0	/* disabled because it interferes the service. */
	/*
	 * If the mount was successfully, and done by root, tell mountd the
	 * good news.  Pid checks are probably unnecessary, but don't hurt.
	 */
	if (rval == 0 && getuid() == 0 &&
	    (mountdfp = fopen(_PATH_MOUNTDPID, "r")) != NULL) {
		int pid;

		if (fscanf(mountdfp, "%d", &pid) == 1 &&
		    pid > 0 && kill(pid, SIGHUP) == -1 && errno != ESRCH)
			err(EXIT_FAILURE, "signal mountd");
		(void)fclose(mountdfp);
	}
#endif

	return rval;
}

int
hasopt(const char *mntopts, const char *option)
{
	int negative, found;
	char *opt, *optbuf;

	if (option[0] == 'n' && option[1] == 'o') {
		negative = 1;
		option += 2;
	} else
		negative = 0;
	optbuf = estrdup(mntopts);
	found = 0;
	for (opt = optbuf; (opt = strtok(opt, ",")) != NULL; opt = NULL) {
		if (opt[0] == 'n' && opt[1] == 'o') {
			if (!strcasecmp(opt + 2, option))
				found = negative;
		} else if (!strcasecmp(opt, option))
			found = !negative;
	}
	free(optbuf);
	return found;
}

static int
mountfs(const char *vfstype, const char *spec, const char *name,
    int flags, const char *options, const char *mntopts,
    int skipmounted, char *buf, size_t buflen)
{
	/* List of directories containing mount_xxx subcommands. */
	static const char *edirs[] = {
#ifdef RESCUEDIR
		RESCUEDIR,
#endif
		_PATH_SBIN,
		_PATH_USRSBIN,
		NULL
	};
	const char ** volatile argv, **edir;
	struct statvfs *sfp, sf;
	pid_t pid;
	int pfd[2];
	int argc, numfs, i, status, maxargc;
	char *optbuf, execname[MAXPATHLEN + 1], execbase[MAXPATHLEN],
	    mntpath[MAXPATHLEN];
	volatile int getargs;

	if (realpath(name, mntpath) == NULL) {
		warn("realpath `%s'", name);
		return 1;
	}

	name = mntpath;

	optbuf = NULL;
	if (mntopts)
		catopt(&optbuf, mntopts);

	if (options) {
		catopt(&optbuf, options);
		getargs = strstr(options, "getargs") != NULL;
	} else
		getargs = 0;

	if (!mntopts && !options)
		catopt(&optbuf, "rw");

	if (getargs == 0 && strcmp(name, "/") == 0 && !hasopt(optbuf, "union"))
		flags |= FLG_UPDATE;
	else if (skipmounted) {
		if ((numfs = getmntinfo(&sfp, MNT_WAIT)) == 0) {
			warn("getmntinfo");
			return 1;
		}
		for(i = 0; i < numfs; i++) {
			const char *mountedtype = sfp[i].f_fstypename;
			size_t cmplen = sizeof(sfp[i].f_fstypename);

			/* remove "puffs|" from comparisons, if present */
#define TYPESIZE (sizeof(PUFFS_TYPEPREFIX)-1)
			if (strncmp(mountedtype,
			    PUFFS_TYPEPREFIX, TYPESIZE) == 0) {
				mountedtype += TYPESIZE;
				cmplen -= TYPESIZE;
			}

			/*
			 * XXX can't check f_mntfromname,
			 * thanks to mfs, union, etc.
			 */
			if (strncmp(name, sfp[i].f_mntonname, MNAMELEN) == 0 &&
			    strncmp(vfstype, mountedtype, cmplen) == 0) {
				if (noise || verbose)
					(void)printf("%s on %s type %.*s: "
					    "%s\n",
					    sfp[i].f_mntfromname,
					    sfp[i].f_mntonname,
					    (int)sizeof(sfp[i].f_fstypename),
					    sfp[i].f_fstypename,
					    "already mounted");
				return 0;
			}
		}
	}
	if (flags & FLG_FORCE)
		catopt(&optbuf, "force");
	if (flags & FLG_RDONLY)
		catopt(&optbuf, "ro");
	/* make -w override -r */
	if (flags & FLG_RDWRITE)
		catopt(&optbuf, "rw");

	if (flags & FLG_UPDATE) {
		catopt(&optbuf, "update");
		/* Figure out the fstype only if we defaulted to ffs */
		if (vfstype == ffs_fstype && statvfs(name, &sf) != -1)
			vfstype = sf.f_fstypename;
	}

	maxargc = 64;
	argv = ecalloc(maxargc, sizeof(*argv));

	if (getargs &&
	    strncmp(vfstype, PUFFS_TYPEPREFIX, sizeof(PUFFS_TYPEPREFIX)-1) == 0)
		(void)snprintf(execbase, sizeof(execbase), "mount_puffs");
	else if (hasopt(optbuf, "rump"))
		(void)snprintf(execbase, sizeof(execbase), "rump_%s", vfstype);
	else
		(void)snprintf(execbase, sizeof(execbase), "mount_%s", vfstype);
	argc = 0;
	argv[argc++] = execbase;
	if (optbuf)
		mangle(optbuf, &argc, &argv, &maxargc);
	argv[argc++] = spec;
	argv[argc++] = name;
	argv[argc] = NULL;

	if ((verbose && buf == NULL) || debug) {
		(void)printf("exec:");
		for (i = 0; i < argc; i++)
			(void)printf(" %s", argv[i]);
		(void)printf("\n");
		if (debug) {
			/*
			 * Don't proceed to attempt to
			 * print fs data, nothing has been
			 * mounted, anything printed would
			 * be for some other filesystem.
			 */
			return 0;
		}
	}

	if (buf) {
		if (pipe(pfd) == -1)
			warn("Cannot create pipe");
	}

	switch (pid = fork()) {
	case -1:				/* Error. */
		warn("fork");
		if (optbuf)
			free(optbuf);
		free(argv);
		return 1;

	case 0:					/* Child. */
		if (debug)
			_exit(0);

		if (buf) {
			(void)close(pfd[0]);
			(void)close(STDOUT_FILENO);
			if (dup2(pfd[1], STDOUT_FILENO) == -1)
				warn("Cannot open fd to mount program");
		}

		/* Go find an executable. */
		edir = edirs;
		do {
			(void)snprintf(execname,
			    sizeof(execname), "%s/%s", *edir, execbase);
			(void)execv(execname, __UNCONST(argv));
			if (errno != ENOENT)
				warn("exec %s for %s", execname, name);
		} while (*++edir != NULL);

		if (errno == ENOENT)
			warn("exec %s for %s: %s", execbase, name, execbase);
		_exit(1);
		/* NOTREACHED */

	default:				/* Parent. */
		if (optbuf)
			free(optbuf);
		free(argv);

		if (buf || getargs) {
			char tbuf[1024], *ptr;
			int nread;

			if (buf == NULL) {
				ptr = tbuf;
				buflen = sizeof(tbuf) - 1;
			} else {
				ptr = buf;
				buflen--;
			}
			(void)close(pfd[1]);
			(void)signal(SIGPIPE, SIG_IGN);
			while ((nread = read(pfd[0], ptr, buflen)) > 0) {
				buflen -= nread;
				ptr += nread;
			}
			*ptr = '\0';
			if (buflen == 0) {
				while (read(pfd[0], &nread, sizeof(nread)) > 0)
					continue;
			}
			if (buf == NULL)
				(void)fprintf(stdout, "%s", tbuf);
		}

		if (waitpid(pid, &status, 0) == -1) {
			warn("waitpid");
			return 1;
		}

		if (WIFEXITED(status)) {
			if (WEXITSTATUS(status) != 0)
				return WEXITSTATUS(status);
		} else if (WIFSIGNALED(status)) {
			warnx("%s: %s", name, strsignal(WTERMSIG(status)));
			return 1;
		}

		if (buf == NULL) {
			if (verbose || noise) {
				if (statvfs(name, &sf) == -1) {
					warn("statvfs %s", name);
					return 1;
				}
				prmount(&sf, 0);
			}
		}
		break;
	}

	return 0;
}

static void
prmount(struct statvfs *sfp, int W)
{
	const struct opt *o;
	struct passwd *pw;
	int flags;
	int f;

	if (strncmp(sfp->f_mntfromname, "/dev/dk", 7) == 0 &&
	    sfp->f_mntfromlabel[0] != '\0') {
		/* a wedge with a label - how would you like it served? */
		switch (W) {
		default:
		case 0:
			(void)printf("%s", sfp->f_mntfromname);
			break;
		case 1:
			(void)printf("NAME=%s", sfp->f_mntfromlabel);
			break;
		case 2:
			(void)printf("NAME=%s (%s)", sfp->f_mntfromlabel,
			    sfp->f_mntfromname);
			break;
		case 3:
			(void)printf("%s (NAME=%s)", sfp->f_mntfromname,
			    sfp->f_mntfromlabel);
			break;
		}
	} else
		(void)printf("%s", sfp->f_mntfromname);

	(void)printf(" on %s type %.*s", sfp->f_mntonname,
	    (int)sizeof(sfp->f_fstypename), sfp->f_fstypename);

	flags = sfp->f_flag & MNT_VISFLAGMASK;
	for (f = 0, o = optnames; flags && o <
	    &optnames[sizeof(optnames)/sizeof(optnames[0])]; o++)
		if (flags & o->o_opt) {
			if (!o->o_silent || verbose)
				(void)printf("%s%s", !f++ ? " (" : ", ",
				    o->o_name);
			flags &= ~o->o_opt;
		}
	if (flags)
		(void)printf("%sunknown flag%s %#x", !f++ ? " (" : ", ",
		    flags & (flags - 1) ? "s" : "", flags);
	if (sfp->f_owner) {
		(void)printf("%smounted by ", !f++ ? " (" : ", ");
		if ((pw = getpwuid(sfp->f_owner)) != NULL)
			(void)printf("%s", pw->pw_name);
		else
			(void)printf("%d", sfp->f_owner);
	}
	if (verbose)
		(void)printf("%sfsid: 0x%x/0x%x",
		    !f++ ? " (" /* ) */: ", ",
		    sfp->f_fsidx.__fsid_val[0], sfp->f_fsidx.__fsid_val[1]);

	if (verbose) {
		(void)printf("%s", !f++ ? " (" : ", ");
		(void)printf("reads: sync %" PRIu64 " async %" PRIu64 "",
		    sfp->f_syncreads, sfp->f_asyncreads);
		(void)printf(", writes: sync %" PRIu64 " async %" PRIu64 "",
		    sfp->f_syncwrites, sfp->f_asyncwrites);
		if (verbose > 1) {
			char buf[2048];

			if (getmntargs(sfp, buf, sizeof(buf)))
				printf(", [%s: %s]", sfp->f_fstypename, buf);
		}
		printf(")\n");
	} else
		(void)printf("%s", f ? ")\n" : "\n");
}

static int
getmntargs(struct statvfs *sfs, char *buf, size_t buflen)
{

	if (mountfs(sfs->f_fstypename, sfs->f_mntfromname, sfs->f_mntonname, 0,
	    "getargs", NULL, 0, buf, buflen))
		return 0;
	else {
		if (*buf == '\0')
			return 0;
		if ((buf = strchr(buf, '\n')) != NULL)
			*buf = '\0';
		return 1;
	}
}

static struct statvfs *
getmntpt(const char *name)
{
	struct statvfs *mntbuf;
	int i, mntsize;

	mntsize = getmntinfo(&mntbuf, MNT_NOWAIT);
	for (i = 0; i < mntsize; i++)
		if (strcmp(mntbuf[i].f_mntfromname, name) == 0 ||
		    strcmp(mntbuf[i].f_mntonname, name) == 0)
			return &mntbuf[i];
	return NULL;
}

static void
catopt(char **sp, const char *o)
{
	char *s, *n;

	s = *sp;
	if (s) {
		easprintf(&n, "%s,%s", s, o);
		free(s);
		s = n;
	} else
		s = estrdup(o);
	*sp = s;
}

static void
mangle(char *options, int *argcp, const char ** volatile *argvp, int *maxargcp)
{
	char *p, *s;
	int argc, maxargc;
	const char **argv, **nargv;

	argc = *argcp;
	argv = *argvp;
	maxargc = *maxargcp;

	for (s = options; (p = strsep(&s, ",")) != NULL;) {
		/* Always leave space for one more argument and the NULL. */
		if (argc >= maxargc - 4) {
			nargv = erealloc(argv, (maxargc << 1) * sizeof(nargv));
			argv = nargv;
			maxargc <<= 1;
		}
		if (*p != '\0') {
			if (*p == '-') {
				argv[argc++] = p;
				p = strchr(p, '=');
				if (p) {
					*p = '\0';
					argv[argc++] = p+1;
				}
			} else {
				argv[argc++] = "-o";
				argv[argc++] = p;
			}
		}
	}

	*argcp = argc;
	*argvp = argv;
	*maxargcp = maxargc;
}

/* Deduce the filesystem type from the disk label. */
static const char *
getfslab(const char *str)
{
	static struct dkwedge_info dkw;
	struct disklabel dl;
	int fd;
	int part;
	const char *vfstype;
	u_char fstype;
	char buf[MAXPATHLEN + 1];
	char *sp, *ep;

	if ((fd = open(str, O_RDONLY)) == -1) {
		/*
		 * Iff we get EBUSY try the raw device. Since mount always uses
		 * the block device we know we are never passed a raw device.
		 */
		if (errno != EBUSY)
			err(EXIT_FAILURE, "cannot open `%s'", str);
		strlcpy(buf, str, MAXPATHLEN);
		if ((sp = strrchr(buf, '/')) != NULL)
			++sp;
		else
			sp = buf;
		for (ep = sp + strlen(sp) + 1;  ep > sp; ep--)
			*ep = *(ep - 1);
		*sp = 'r';

		/* Silently fail here - mount call can display error */
		if ((fd = open(buf, O_RDONLY)) == -1)
			return NULL;
	}

	/* Check to see if this is a wedge. */
	if (ioctl(fd, DIOCGWEDGEINFO, &dkw) == 0) {
		/* Yup, this is easy. */
		close(fd);
		if (*dkw.dkw_ptype)
			return dkw.dkw_ptype;
		return NULL;
	}

	if (ioctl(fd, DIOCGDINFO, &dl) == -1) {
		(void) close(fd);
		return NULL;
	}

	(void) close(fd);

	part = str[strlen(str) - 1] - 'a';

	if (part < 0 || part >= dl.d_npartitions)
		return NULL;

	/* Return NULL for unknown types - caller can fall back to ffs */
	if ((fstype = dl.d_partitions[part].p_fstype) >= FSMAXMOUNTNAMES)
		vfstype = NULL;
	else
		vfstype = mountnames[fstype];

	return vfstype;
}

static void
usage(void)
{

	(void)fprintf(stderr, "Usage: %s [-Aadfruvw] [-t type]\n"
	    "\t%s [-dfruvw] special | node\n"
	    "\t%s [-dfruvw] [-o options] [-t type] special node\n",
	    getprogname(), getprogname(), getprogname());
	exit(1);
}
